rocAL/rocAL/source/sequence_file_source_reader.cpp

```diff
@@ -0,0 +1,306 @@
+    {
+        _last_id.erase(0, last_slash_idx + 1);
+    }
+
+    _current_fPtr = fopen(file_path.c_str(), "rb"); // Open the file,
+
+    if (!_current_fPtr) // Check if it is ready for reading
+        return 0;
+
+    fseek(_current_fPtr, 0, SEEK_END); // Take the file read pointer to the end
+
+    _current_file_size = ftell(_current_fPtr); // Check how many bytes are there between and the current read pointer position (end of the file)
+
+    if (_current_file_size == 0)
+    { // If file is empty continue
+        fclose(_current_fPtr);
+        _current_fPtr = nullptr;
+        return 0;
+    }
+
+    fseek(_current_fPtr, 0, SEEK_SET); // Take the file pointer back to the start
+
+    return _current_file_size;
+}
+
+size_t SequenceFileSourceReader::read(unsigned char *buf, size_t read_size)
+{
+    if (!_current_fPtr)
+        return 0;
+
+    // Requested read size bigger than the file size? just read as many bytes as the file size
+    read_size = (read_size > _current_file_size) ? _current_file_size : read_size;
+
+    size_t actual_read_size = fread(buf, sizeof(unsigned char), read_size, _current_fPtr);
+    return actual_read_size;
+}
+
+int SequenceFileSourceReader::close()
+{
+    return release();
+}
+
+SequenceFileSourceReader::~SequenceFileSourceReader()
+{
+    release();
+}
+
+int SequenceFileSourceReader::release()
+{
+    if (!_current_fPtr)
+        return 0;
+    fclose(_current_fPtr);
+    _current_fPtr = nullptr;
+    return 0;
+}
+
+void SequenceFileSourceReader::reset()
+{
+    _shuffle_time.start();
+    if (_shuffle)
+        std::random_shuffle(_sequence_frame_names.begin(), _sequence_frame_names.end());
+    _shuffle_time.end();
+    _read_counter = 0;
+    _curr_file_idx = 0;
+}
+
+Reader::Status SequenceFileSourceReader::get_sequences()
+{
+    for (unsigned folder_idx = 0; folder_idx < _folder_file_names.size(); folder_idx++)
+    {
+        if(_folder_file_names[folder_idx].size() == 0)
+        {
+            WRN("\nFolder #" + TOSTR(folder_idx) + "does not have any files")
+            continue;
+        }
+        if (_sequence_length > _folder_file_names[folder_idx].size())
+        {
+            THROW("Sequence length is not valid");
+        }
+        for (unsigned file_idx = 0; (file_idx + (_stride * (_sequence_length - 1))) < _folder_file_names[folder_idx].size(); file_idx+=_step)
+        {
+            if (get_sequence_shard_id() != _shard_id)
+            {
+                _sequence_count_all_shards++;
+                incremenet_sequence_id();
+                continue;
+            }
+            _in_batch_read_count++;
+            _in_batch_read_count = (_in_batch_read_count % _batch_count == 0) ? 0 : _in_batch_read_count;
+
+            std::vector<std::string> temp;
+            for (unsigned frame_count = 0, frame_idx = file_idx; (frame_count < _sequence_length) ; frame_count++, frame_idx+=_stride)
+            {
+                temp.push_back(_folder_file_names[folder_idx][frame_idx]);
```

Comment:
instead of using temp, push into sequence_frame_names to avoid extra copy

Reply:
Temp is used here to keep track of the last sequence that is obtained. And the last sequence will be later used for padding